plugin/arcconf/arcconf.py

```diff
@@ -0,0 +1,657 @@
+
+
+class Arcconf(IPlugin):
+    _DEFAULT_BIN_PATHS = [
+        "/usr/bin/arcconf",
+        "/usr/sbin/arcconf"]
+
+    def __init__(self):
+        self._arcconf_bin = None
+        self._tmo_ms = 30000
+
+    def _find_arcconf(self):
+        """
+        Try _DEFAULT_MDADM_BIN_PATHS
+        """
+        for cur_path in Arcconf._DEFAULT_BIN_PATHS:
+            if os.path.lexists(cur_path):
+                self._arcconf_bin = cur_path
+
+        if not self._arcconf_bin:
+            raise LsmError(
+                ErrorNumber.INVALID_ARGUMENT,
+                "arcconf is not installed correctly")
+
+    @_handle_errors
+    def plugin_register(self, uri, password, timeout, flags=Client.FLAG_RSVD):
+        if os.geteuid() != 0:
+            raise LsmError(
+                ErrorNumber.INVALID_ARGUMENT,
+                "This plugin requires root privilege both daemon and client")
+        uri_parsed = uri_parse(uri)
+        self._arcconf_bin = uri_parsed.get('parameters', {}).get('arcconf')
+        if not self._arcconf_bin:
+            self._find_arcconf()
+
+    @_handle_errors
+    def plugin_unregister(self, flags=Client.FLAG_RSVD):
+        pass
+
+    @_handle_errors
+    def job_status(self, job_id, flags=0):
+        return (JobStatus.COMPLETE, 100, None)
+
+    @_handle_errors
+    def job_free(self, job_id, flags=Client.FLAG_RSVD):
+        pass
+
+    @_handle_errors
+    def plugin_info(self, flags=Client.FLAG_RSVD):
+        return "Arcconf Plugin", VERSION
+
+    @_handle_errors
+    def time_out_set(self, ms, flags=Client.FLAG_RSVD):
+        self._tmo_ms = ms
+
+    @_handle_errors
+    def time_out_get(self, flags=Client.FLAG_RSVD):
+        return self._tmo_ms
+
+    @_handle_errors
+    def capabilities(self, system, flags=Client.FLAG_RSVD):
+        cur_lsm_syss = self.systems()
+        if system.id not in list(s.id for s in cur_lsm_syss):
+            raise LsmError(
+                ErrorNumber.NOT_FOUND_SYSTEM,
+                "System not found")
+        cap = Capabilities()
+        cap.set(Capabilities.DISKS)
+        cap.set(Capabilities.SYS_FW_VERSION_GET)
+        cap.set(Capabilities.SYS_MODE_GET)
+        cap.set(Capabilities.DISK_LOCATION)
+        cap.set(Capabilities.DISK_VPD83_GET)
+        if system.mode != System.MODE_HBA:
+            cap.set(Capabilities.VOLUMES)
+            cap.set(Capabilities.VOLUME_RAID_CREATE)
+            cap.set(Capabilities.VOLUME_DELETE)
+
+        return cap
+
+    def _arcconf_exec(self, arcconf_cmds, flag_force=False):
+        arcconf_cmds.insert(0, self._arcconf_bin)
+        if flag_force:
+            arcconf_cmds.append('noprompt')
+        try:
+            output = cmd_exec(arcconf_cmds)
+        except OSError as os_error:
+            if os_error.errno == errno.ENOENT:
+                raise LsmError(
+                    ErrorNumber.INVALID_ARGUMENT,
+                    "arcconf binary '%s' does not exist." %
+                    self._arcconf_bin)
```

Comment:
Does this mean we need to try executing a command like : self._arcconf_exec(['list']) or something like that to verify _arcconf_bin actually exists?